compat/mingw.c

```diff
@@ -759,7 +801,7 @@ int mingw_open (const char *filename, int oflags, ...)
 		return -1;
 	}
 
-	if ((oflags & O_APPEND) && !is_local_named_pipe_path(filename))
+	if ((oflags & O_APPEND) && is_local_disk_file(filename))
```

Comment:
Previously, this condition triggered when appending to anything but a local named pipe. Now, this is restricted to local disk files? That sounds a bit overzealous.

It would probably make more sense to trigger the new (expensive!) code path _only_ if the `open_fn()` call further down in this function failed (essentially: try running with `mingw_open_append()` first, and re-try using `_wopen()` only if it failed, and only if `GetLastError()` has the expected value, which you can easily find out by printing it when the function call fails in your scenario, and then only if that `GetDriveTypeW()` indicates a remote path).

Of course, an alternative would be to simply ask `has_dos_drive_prefix()` as an indicator for a local file:

```c
if ((oflags & O_APPEND) && !is_local_named_pipe_path(filename) && has_dos_drive_prefix(filename))
```

That should be good-enough in your use case, and fix the issue reliably without having to go through expensive Win32 API calls.

Reply:
After re-checking git commit(https://github.com/git-for-windows/git/commit/d641097589160eb795127d8dbcb14c877c217b60) and MSDN doc(https://docs.microsoft.com/en-us/windows/win32/fileio/file-access-rights-constants), "mingw_open_append" is introduced for atomatic append, it uses "FILE_APPEND_DATA" only flag to achieve this, and according to the MSDN doc, it should only work for local files: 
<html>
<body>
<!--StartFragment-->
<i>
For a file object, the right to append data to the file. (For local files, write operations will not overwrite existing data if this flag is specified without FILE_WRITE_DATA.) For a directory object, the right to create a subdirectory (FILE_ADD_SUBDIRECTORY).
</i>
<br class="Apple-interchange-newline">
<!--EndFragment-->
</body>
</html>

Reply:
> `mingw_open_append` will still succeed on a UNC path with `FILE_APPEND_DATA`, but it will fail in later `write` calls.

What is the error?

Can you reproduce the same error when using the "administrative share" `\\localhost\c$`?

Reply:
thank you for your feedback.
`GetLastError()` after `write` returns: `ERROR_INVALID_PARAMETER : The parameter is incorrect.`
and I failed to reproduce it when using `\\localhost\c$` , it creates and writes file successfully.

Reply:
> `GetLastError()` after `write` returns: `ERROR_INVALID_PARAMETER : The parameter is incorrect.`

Okay. Unfortunately, this is not really helpful.

The problem I see here is the adverse effects of this PR for the common case: while `is_local_named_pipe_path()` only looks at the string and is therefore very fast, the new `is_local_disk_file()` function calls not only `GetFullPathNameW()` but also `GetDriveTypeW()`, i.e. two I/O operations, read: comparatively expensive functions.

Granted, this is "only" when we're opening a file with the intention to append to it. But still, it is a complication, and judging by the number of bug reports we received, the fix is actually not necessary for all network shares, therefore the workaround is a bit overzealous.

How about doing something different and going for a config option instead? Something like this:

```c
static int append_atomically = -1;

[...]

if (append_atomically < 0 &&
    git_config_get_bool("windows.appendatomically", &append_atomically))
        append_atomically = 1; /* default to `true` */

if (append_atomically &&
    (oflags & O_APPEND) && !is_local_named_pipe_path(filename))
        open_fn = mingw_open_append;
[...]
```

We could _then_ detect the `ERROR_INVALID_PARAMETER` condition and suggest to disable atomic appending via the config option.

That would get us mostly to the best of both worlds: best performance for the common case, and a way out of the problematic scenario via an opt-in config setting.

What do you think?

Reply:
I agree with you. I'll update it in one or two weeks. Sorry for busy these days.

Reply:
Thank you!